daemon/mgr/network.go

```diff
@@ -744,3 +744,63 @@ func joinOptions(endpoint *types.Endpoint) ([]libnetwork.EndpointOption, error)
 	joinOptions = append(joinOptions, libnetwork.JoinOptionPriority(nil, endpoint.Priority))
 	return joinOptions, nil
 }
+
+// getSandboxPortMapInfo retrieves the current port-mapping programmed for the given sandbox.
+func getSandboxPortMapInfo(sb libnetwork.Sandbox) apitypes.PortMap {
+	pm := apitypes.PortMap{}
+	if sb == nil {
+		return pm
+	}
+
+	for _, ep := range sb.Endpoints() {
+		pm, _ = getEndpointPortMapInfo(ep)
+		if len(pm) > 0 {
+			break
+		}
+	}
+	return pm
+}
+
+func getEndpointPortMapInfo(ep libnetwork.Endpoint) (apitypes.PortMap, error) {
+	pm := apitypes.PortMap{}
+	driverInfo, err := ep.DriverInfo()
+	if err != nil {
+		return pm, err
+	}
+
+	if driverInfo == nil {
+		return pm, nil
+	}
+
+	// get exposedPorts from driverInfo, which was open by --expose and so on
+	if expData, ok := driverInfo[netlabel.ExposedPorts]; ok {
+		if exposedPorts, ok := expData.([]networktypes.TransportPort); ok {
+			for _, tp := range exposedPorts {
+				natPort, err := nat.NewPort(tp.Proto.String(), strconv.Itoa(int(tp.Port)))
+				if err != nil {
+					return pm, fmt.Errorf("failed to parse Port value(%v):%v", tp.Port, err)
+				}
+				pm[string(natPort)] = nil
+			}
+		}
+	}
+
+	mapData, ok := driverInfo[netlabel.PortMap]
+	if !ok {
+		return pm, nil
+	}
+
+	// get the port-mapping from driverInfo, which was open by -p HostPort:port
+	if portMapping, ok := mapData.([]networktypes.PortBinding); ok {
```

Comment:
You could use https://github.com/alibaba/pouch/blob/master/docs/contributions/code_styles.md#rule006---return-fast-to-indent-less to indent less.

```
portMapping, ok := mapData.([]networktypes.PortBinding)
if !ok {
    return pm, nil
}
for _, pp := range portMapping {
      ......
}
return pm, nil
```